File: src/services/discourseSSOService.js
import crypto from 'crypto';
import { config } from '../config';

/**
 * Generate Discourse SSO payload for a user
 * @param {Object} user - User object
 * @param {string} sso - Base64 encoded SSO payload from Discourse
 * @param {string} sig - Signature from Discourse
 * @returns {string} URL-encoded payload for Discourse SSO
 */
export function generateDiscourseSSOPayload(user, sso, sig) {
  // Verify the signature from Discourse
  const hmac = crypto.createHmac('sha256', config.DISCOURSE_SSO_SECRET);
  hmac.update(sso);
  const computedSig = hmac.digest('hex');
  
  if (computedSig !== sig) {
    throw new Error('Invalid Discourse SSO signature');
  }
  
  // Decode the payload
  const payload = Buffer.from(sso, 'base64').toString('utf-8');
  const nonce = new URLSearchParams(payload).get('nonce');
  
  // Prepare response payload
  const responseParams = {
    nonce: nonce,
    external_id: user.id.toString(),
    email: user.email,
    username: user.name.replace(/\s+/g, '_').toLowerCase(),
    name: user.name,
    // Additional optional parameters
    groups: user.roles.join(',')
  };
  
  // Encode the response
  const returnPayload = new URLSearchParams(responseParams).toString();
  const base64Payload = Buffer.from(returnPayload).toString('base64');
  
  // Generate signature for the response
  const returnHmac = crypto.createHmac('sha256', config.DISCOURSE_SSO_SECRET);
  returnHmac.update(base64Payload);
  const returnSig = returnHmac.digest('hex');
  
  return `sso=${encodeURIComponent(base64Payload)}&sig=${returnSig}`;
}

File: src/services/discourse_sso_service.rs
use crate::config::Config;
use base64::{Engine as _, engine::general_purpose};
use hmac::{Hmac, Mac};
use sha2::Sha256;
use url::form_urlencoded;
use std::collections::HashMap;
use std::error::Error;

type HmacSha256 = Hmac<Sha256>;

/// Error type for Discourse SSO operations
#[derive(Debug, thiserror::Error)]
pub enum DiscourseSSOError {
    #[error("Invalid SSO signature")]
    InvalidSignature,
    #[error("Failed to decode SSO payload: {0}")]
    DecodingError(String),
    #[error("Missing nonce in SSO payload")]
    MissingNonce,
    #[error("HMAC error: {0}")]
    HmacError(String),
    #[error("Encoding error: {0}")]
    EncodingError(String),
}

/// Generate Discourse SSO payload for a user
///
/// # Arguments
///
/// * `user` - User object
/// * `sso` - Base64 encoded SSO payload from Discourse
/// * `sig` - Signature from Discourse
/// * `config` - Application configuration
///
/// # Returns
///
/// * `Result<String, DiscourseSSOError>` - URL-encoded payload for Discourse SSO
pub fn generate_discourse_sso_payload(
    user: &impl UserInfo,
    sso: &str,
    sig: &str,
    config: &Config,
) -> Result<String, DiscourseSSOError> {
    // Verify the signature from Discourse
    let mut mac = HmacSha256::new_from_slice(config.discourse_sso_secret.as_bytes())
        .map_err(|e| DiscourseSSOError::HmacError(e.to_string()))?;
    
    mac.update(sso.as_bytes());
    let computed_sig = hex::encode(mac.finalize().into_bytes());
    
    if computed_sig != sig {
        return Err(DiscourseSSOError::InvalidSignature);
    }
    
    // Decode the payload
    let decoded_payload = general_purpose::STANDARD
        .decode(sso)
        .map_err(|e| DiscourseSSOError::DecodingError(e.to_string()))?;
    
    let payload_str = String::from_utf8(decoded_payload)
        .map_err(|e| DiscourseSSOError::DecodingError(e.to_string()))?;
    
    // Parse the payload to get the nonce
    let params: HashMap<String, String> = form_urlencoded::parse(payload_str.as_bytes())
        .into_owned()
        .collect();
    
    let nonce = params.get("nonce")
        .ok_or(DiscourseSSOError::MissingNonce)?;
    
    // Prepare response payload
    let mut response_params = HashMap::new();
    response_params.insert("nonce", nonce.as_str());
    response_params.insert("external_id", &user.id().to_string());
    response_params.insert("email", user.email());
    response_params.insert("username", &user.name().replace(" ", "_").to_lowercase());
    response_params.insert("name", user.name());
    
    // Additional optional parameters
    if let Some(roles) = user.roles() {
        response_params.insert("groups", &roles.join(","));
    }
    
    // Encode the response
    let return_payload = form_urlencoded::Serializer::new(String::new())
        .extend_pairs(response_params.iter())
        .finish();
    
    let base64_payload = general_purpose::STANDARD.encode(return_payload.as_bytes());
    
    // Generate signature for the response
    let mut return_mac = HmacSha256::new_from_slice(config.discourse_sso_secret.as_bytes())
        .map_err(|e| DiscourseSSOError::HmacError(e.to_string()))?;
    
    return_mac.update(base64_payload.as_bytes());
    let return_sig = hex::encode(return_mac.finalize().into_bytes());
    
    // Combine and return final payload
    Ok(format!("sso={}&sig={}", urlencoding::encode(&base64_payload), return_sig))
}

/// Trait for user information needed for SSO
pub trait UserInfo {
    fn id(&self) -> &str;
    fn email(&self) -> &str;
    fn name(&self) -> &str;
    fn roles(&self) -> Option<&Vec<String>>;
}
